File: gui_view/src/main/java/com/goexp/galgame/gui/view/game/part/StateChangeController.java
package com.goexp.galgame.gui.view.game.part;

import com.goexp.galgame.common.model.GameState;
import com.goexp.galgame.gui.model.Game;
import com.goexp.galgame.gui.task.game.ChangeGameTask;
import com.goexp.galgame.gui.view.DefaultController;
import com.goexp.galgame.gui.task.TaskService;
import javafx.beans.value.ChangeListener;
import javafx.collections.FXCollections;
import javafx.concurrent.Service;
import javafx.fxml.FXML;
import javafx.scene.control.ChoiceBox;
import javafx.util.StringConverter;

import java.util.Comparator;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class StateChangeController extends DefaultController {

    @FXML
    private ChoiceBox<GameState> choiceState;

    private ChangeListener<GameState> listener;
    private Game targetGame;
    private Service<Void> changeGameStateService = new TaskService<>(() -> new ChangeGameTask.Like(targetGame));


    protected void initialize() {
        var types = Stream.of(
                GameState.values()
        )
                .sorted(Comparator.reverseOrder())
                .collect(Collectors.toList());

        choiceState.setConverter(new StringConverter<>() {
            public String toString(GameState gameState) {
                return gameState.name;
            }

            public GameState fromString(String string) {
                return GameState.from(string);
            }
        });

        choiceState.setItems(FXCollections.observableArrayList(types));


        listener = (o, oldValue, newValue) -> {

            if (newValue != null) {
                logger.debug("<Action>Value:{},New:{}", choiceState.getValue(), newValue);

                targetGame.state.set(newValue);

                changeGameStateService.restart();
            }
        };
    }

    public void load(Game game) {
        choiceState.valueProperty().removeListener(listener);
        choiceState.setValue(game.state.get());
        targetGame = game;
        choiceState.valueProperty().addListener(listener);
    }
}

File: gui_view/src/main/scala/com/goexp/galgame/gui/view/game/part/StateChangeController.scala
package com.goexp.galgame.gui.view.game.part

import com.goexp.galgame.common.model.GameState
import com.goexp.galgame.gui.model.Game
import com.goexp.galgame.gui.task.TaskService
import com.goexp.galgame.gui.task.game.change.Like
import com.goexp.galgame.gui.view.DefaultController
import javafx.beans.value.ChangeListener
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.control.ChoiceBox
import javafx.util.StringConverter

import scala.jdk.CollectionConverters._

class StateChangeController extends DefaultController {
  @FXML private var choiceState: ChoiceBox[GameState] = _

  private var targetGame: Game = _
  final private val changeGameStateService = TaskService(() => new Like(targetGame))

  private val listener: ChangeListener[GameState] = (_, _, newValue) => {
    if (newValue != null) {
      logger.debug("<Action>Value:{},New:{}", choiceState.getValue, newValue)
      targetGame.state.set(newValue)
      changeGameStateService.restart()
    }
  }

  override protected def initialize() = {

    choiceState.setConverter(new StringConverter[GameState]() {
      override def toString(gameState: GameState) = gameState.name

      override def fromString(string: String) = GameState.from(string)
    })

    val types = GameState.values.to(LazyList).sortBy(gs => gs.value).reverse.asJava
    choiceState.setItems(FXCollections.observableArrayList(types))
  }

  def load(game: Game) = {

    choiceState.valueProperty.removeListener(listener)
    choiceState.setValue(game.state.get)
    targetGame = game
    choiceState.valueProperty.addListener(listener)
  }
}